### PR TITLE
Support all `provider_installation` methods with provider cache

### LINKED ISSUE #3117: Take into account user `provider_installation` settings when using Terragrunt Provider Cache
## Describe the feature

1. If there is a `provider_installation` block in the use CLI configuration `.terraformrc`, `terraform` returns errors `Error: At most one provider_installation block can be specified`, this is because Terragrunt Provider Cache adds another block, so we need to merge them into one.
2. If the user CLI configuration `.terraformrc` has `network_mirror` settings, this should be taken into account when configuring **Terragrunt Provider Cache**. That is, **Terragrunt Provider Cache** server should make requests not to the remote register, but to the one specified in the `network_mirror` settings.

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

Fixes #3117.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

File: terraform/cache/models/provider.go [hunk 1]
```diff
@@ -2,15 +2,92 @@ package models
...
+	Protocols []string    `json:"protocols"`
+	Platforms []*Platform `json:"platforms"`
+}
+
+type Platform struct {
+	OS   string `json:"os"`
+	Arch string `json:"arch"`
+}
+
+func ParsePlatform(str string) *Platform {
```
Comment (denis256) [inline terraform/cache/models/provider.go:61]: `ParsePlatform` seems to be unused
```
$ grep -irn "ParsePlatform" .
./terraform/cache/models/provider.go:61:func ParsePlatform(str string) *Platform {

$
```

File: terraform/cache/handlers/helper.go [hunk 2]
```diff
@@ -2,26 +2,65 @@ package handlers
...
+	"strconv"
 )
 
+func ResponseReader(resp *http.Response) (io.ReadCloser, error) {
+	// Check that the server actually sent compressed data
+	switch resp.Header.Get("Content-Encoding") {
+	case "gzip":
+		reader, err := gzip.NewReader(resp.Body)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply terraform/cache/handlers/helper.go:18]: No, the body is closed with this code

https://github.com/golang/go/blob/01ad44bc08c7ea3b8a7d1d989051046c77b6a63d/src/net/http/httputil/reverseproxy.go#L324-L327

[the remark this answers is not in the document]
Comment (levkohimins) [reply terraform/cache/handlers/helper.go:18]: But I found a place where it was missed.

https://github.com/gruntwork-io/terragrunt/blob/f2586b969b6b6dd0192fc0ef52b3a4e900b0447d/terraform/cache/handlers/provider_network_mirror.go#L119

File: terraform/cache/models/provider.go (hunk 1 above)
Comment (levkohimins) [reply terraform/cache/models/provider.go]: Indeed. Thanks.

File: terraform/cache/handlers/provider.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+type ProviderHandler interface {
+	// CanHandleProvider returns true if the given provider can be handled by this handler.
+	CanHandleProvider(provider *models.Provider) bool
+
+	// GetVersions serves a request that returns all versions for a single provider.
+	GetVersions(ctx echo.Context, provider *models.Provider) error
+
+	// GetPlatfrom serves a request that returns a provider for a specific platform.
+	GetPlatfrom(ctx echo.Context, provider *models.Provider, downloaderController router.Controller, cacheRequestID string) error
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply terraform/cache/handlers/provider.go:35]: I’m not sure if I follow. Could you please explain your suggestion?

**Comment by denis256** [comment]
I think lint rules/exclusions should be updated

```
INFO [runner] linters took 41.599419775s with stages: goanalysis_metalinter: 41.592211209s 
terraform/cache/handlers/provider_network_mirror.go:119:23: Error return value of `resp.Body.Close` is not checked (errcheck)
        defer resp.Body.Close()
```

**Comment by levkohimins** [comment]
> I think lint rules/exclusions should be updated
> 
> ```
> INFO [runner] linters took 41.599419775s with stages: goanalysis_metalinter: 41.592211209s 
> terraform/cache/handlers/provider_network_mirror.go:119:23: Error return value of `resp.Body.Close` is not checked (errcheck)
>         defer resp.Body.Close()
> ```

Fixed

File: terraform/cache/handlers/provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (levkohimins) [reply terraform/cache/handlers/provider.go]: Fixed

**Review by denis256 — APPROVED** [review]

**Comment by levkohimins** [comment]
Resolved in [v0.58.7](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.58.7) release.

**Pull request merged**